Objects/dictobject.c

```diff
@@ -141,6 +142,23 @@ To avoid slowing down lookups on a near-full table, we resize the table when
 it's USABLE_FRACTION (currently two-thirds) full.
 */
 
+#ifdef Py_GIL_DISABLED
+
+static inline void
+ASSERT_DICT_LOCKED(PyObject *op)
+{
+    if (Py_REFCNT(op) != 1) {
+        _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(op);
+    }
```

Comment:
We should push the refcount check down into the assert. We'll want this in general, not just for dict. Additionally, we want the refcount check  `Py_REF_DEBUG` macro. It contains atomic loads that won't be optimized away (in theory the could, but current GCC and Clang versions do not as far as I can tell.)